Sources/RabbitMq/Connection.swift

```diff
@@ -1,46 +1,149 @@
 import AMQPClient
+import AsyncAlgorithms
 import Foundation
+import Logging
 import NIO
 import NIOSSL
+import ServiceLifecycle
+
+let PollingConnectionSleepInterval = Duration.milliseconds(100)
 
 public actor Connection {
-    private let config: AMQPConnectionConfiguration
+    private let url: String
     private let eventLoop: EventLoop
+    private let config: AMQPConnectionConfiguration
+    let logger: Logger  // shared to users of Connection
 
     private var channel: AMQPChannel?
     private var connection: AMQPConnection?
+    private let newConsumers: AsyncChannel<RetryingConsumer>
+
+    private var connecting = false
+
+    public var isConnected: Bool {
+        if let conn = self.connection {
+            return conn.isConnected
+        }
+        return false
+    }
 
     public init(
-        eventLoop: EventLoop = MultiThreadedEventLoopGroup(numberOfThreads: 1).next(),
         _ url: String = "",
-        tls: TLSConfiguration = TLSConfiguration.makeClientConfiguration()
+        eventLoop: EventLoop = MultiThreadedEventLoopGroup(numberOfThreads: 1).next(),
+        tls: TLSConfiguration = TLSConfiguration.makeClientConfiguration(),
+        logger: Logger = Logger(label: "\(Connection.self)")
     ) throws {
+        self.url = url
         self.eventLoop = eventLoop
         self.config = try AMQPConnectionConfiguration.init(url: url, tls: tls)
+        self.logger = logger
+
+        self.newConsumers = AsyncChannel<RetryingConsumer>()
     }
 
-    public func isConnected() -> Bool {
-        if let conn = self.connection {
-            return conn.isConnected
+    // Internal use only, for retrying consumers functionality
+    func addRetryingConsumer(consumer: RetryingConsumer) async {
+        await newConsumers.send(consumer)
```

Comment:
I just realized that if you're trying to add consumers before the run function is executed, this could hang. That might be a problem.

Reply:
I think I've seen this...